internal/config/cgroupmanager/supported.go

```diff
@@ -0,0 +1,123 @@
+// +build linux
+
+package cgroupmanager
+
+import (
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+
+	libpodcgroups "github.com/containers/libpod/pkg/cgroups"
+	libctrcgroups "github.com/opencontainers/runc/libcontainer/cgroups"
+	"github.com/pkg/errors"
+)
+
+var (
+	_cgroupHasHugetlbOnce sync.Once
+	_cgroupHasHugetlb     bool
+	_cgroupHasHugetlbErr  error
+
+	_cgroupHasPidOnce sync.Once
+	_cgroupHasPid     bool
+	_cgroupHasPidErr  error
+
+	_cgroupHasMemorySwapOnce sync.Once
+	_cgroupHasMemorySwap     bool
+	_cgroupHasMemorySwapErr  error
+
+	_cgroupIsV2Once sync.Once
+	_cgroupIsV2     bool
+	_cgroupIsV2Err  error
+)
+
+// initializeCgroups initializes all of the singleton variables
+// that store the node's cgroup configuration. Currently, we check
+// hugetlb, cgroup v1 or v2, pid and memory swap support
+// We check the error at server initialization, and if we error, shutdown
+// cri-o early, instead of when we're already trying to run containers.
+func initializeCgroups() error {
+	toInit := []struct {
+		initializer func() bool
+		err         *error
+	}{
+		{
+			initializer: CgroupHasHugetlb,
+			err:         &_cgroupHasHugetlbErr,
+		},
+		{
+			initializer: CgroupHasPid,
+			err:         &_cgroupHasPidErr,
+		},
+		{
+			initializer: CgroupIsV2,
+			err:         &_cgroupIsV2Err,
+		},
+		{
+			initializer: CgroupHasMemorySwap,
+			err:         &_cgroupHasMemorySwapErr,
+		},
+	}
+	for _, i := range toInit {
+		i.initializer()
+		if *i.err != nil {
+			return errors.Wrapf(*i.err, "failed to initialize cgroup information")
+		}
+	}
+	return nil
+}
+
+// CgroupHasHugetlb returns whether the hugetlb controller is present
+func CgroupHasHugetlb() bool {
+	_cgroupHasHugetlbOnce.Do(func() {
+		if CgroupIsV2() {
+			if _, err := ioutil.ReadDir("/sys/fs/cgroup/hugetlb"); err != nil {
+				_cgroupHasHugetlbErr = errors.Wrap(err, "readdir /sys/fs/cgroup/hugetlb")
+				return
+			}
+			_cgroupHasHugetlb = true
+			return
+		}
+
+		controllers, err := ioutil.ReadFile("/sys/fs/cgroup/cgroup.controllers")
+		if err != nil {
+			_cgroupHasHugetlbErr = errors.Wrap(err, "read /sys/fs/cgroup/cgroup.controllers")
+			return
+		}
+		_cgroupHasHugetlb = strings.Contains(string(controllers), "hugetlb")
+	})
+	return _cgroupHasHugetlb
+}
+
+// CgroupHasPid returns whether the pid controller is present
+func CgroupHasPid() bool {
+	_cgroupHasPidOnce.Do(func() {
+		_, err := libctrcgroups.FindCgroupMountpoint("", "pids")
```

Comment:
This is a super ineffective way to check, as it involves parsing /proc/self/mountinfo.

I think the whole section of checking for controllers need to be rewritten. For v2, we check the `cgroup.controllers` file. For v1, unless you have a weird setup, checking for a particular directory presence is fine.

IOW maybe generalize what we have for hugetlb, check for all controllers at once, cache the result, and return it on demand.

Reply:
I would do:

```go
_availableControllers map string[bool]
```

```
func CgroupHasController(c string) bool {
  has, ok := _availableControllers[c]
  if !ok {
    logrus.Warnf("CgroupHasController: invalid argument %s", c)
  }
  return has
}
```

(plus locking for the map)

Reply:
I've gone for a slightly different solution that doesn't need a lock nor a map (which I think its a bit bloated if we only need two bools)